File: ios/Shared/WidgetSnapshot.swift
import Foundation

let widgetAppGroup = "group.com.ilovetrains.ios"
let homeWidgetKind = "ILoveTrainsHome"
let widgetSnapshotFileName = "widget-v1.json"
let widgetScheduleHours = 168
let widgetBoardLimit = 8
let widgetLiveRefresh: Millis = 900_000
let widgetTimelineHorizon: Millis = 10_800_000
let widgetHomeURL = URL(string: "ilovetrains://home")!
let widgetSetupURL = URL(string: "ilovetrains://setup")!
private let hourMillis: Millis = 3_600_000
private let widgetCalendar = sydneyCalendar

struct WidgetStop: Codable, Equatable, Sendable {
    var id: String
    var name: String
    var modes: [String]

    var station: Station { Station(id: id, name: name, modes: Set(modes)) }
}

struct WidgetTrip: Codable, Equatable, Sendable {
    var id: String
    var from: WidgetStop
    var to: WidgetStop
}

struct WidgetScheduleEntry: Codable, Equatable, Sendable {
    var at: Millis
    var tripId: String
    var reverse: Bool
}

struct WidgetFocus: Codable, Equatable, Sendable {
    var tripId: String
    var reverse: Bool
    var pinned: Bool
    var journey: Journey
    var board: BoardData
    var expiresAt: Millis
}

/// Everything the widget may know about the rider; only the app writes it.
struct WidgetSnapshot: Codable, Equatable, Sendable {
    var schemaVersion = 1
    var writtenAt: Millis
    var trips: [WidgetTrip]
    var schedule: [WidgetScheduleEntry]
    var focus: WidgetFocus?
    var modes: [String]
    var transferCap: Int?
    var boards: [BoardData]

    func board(from: String, to: String) -> BoardData? {
        boards.first { $0.from.id == from && $0.to.id == to }
    }

    func eligible(_ journey: Journey) -> Bool {
        journeyAllowed(journey, modes: Set(modes)) && (transferCap.map { journey.legs.count - 1 <= $0 } ?? true)
    }
}

struct WidgetAnswer: Equatable, Sendable {
    var trip: WidgetTrip
    var reverse: Bool
    var focus: WidgetFocus?

    var from: WidgetStop { reverse ? trip.to : trip.from }
    var to: WidgetStop { reverse ? trip.from : trip.to }
}

struct WidgetRequest: Equatable, Sendable {
    var key: String
    var from: Station
    var to: Station
    var modes: Set<String>
    var at: Millis?
    var transferLimit: Int?
    var focusJourneyKey: String?
    var fallback: BoardData?
}

struct WidgetContent: Equatable, Sendable {
    var date: Millis
    var answer: WidgetAnswer?
    var board: BoardData?
    /// The header's answer: the followed service, or the recommendation among the board's eligible departures.
    var lead: Journey?
    /// The cancelled departure the lead stands in for, as the header's `<time> cancelled · next train`.
    var replaced: Journey?
    var following: [Journey] = []
    var freshness: WidgetFreshness?

    /// A board, as the medium prints it: the replaced departure, the lead, then what leaves after it.
    var rows: [Journey] { Array(([replaced, lead].compactMap { $0 } + following).prefix(3)) }
}

enum WidgetFreshness: Equatable, Sendable {
    case updated(Millis), lastKnown(Millis), offline(Millis), timetable, unavailable

    /// A widget is almost never live, so it always prints the clock time of its data, never an age that freezes.
    var text: String {
        switch self {
        case let .updated(at): "Last updated \(clockTime(at))"
        case let .lastKnown(at): "Last known · Last updated \(clockTime(at))"
        case let .offline(at): "Offline · Last updated \(clockTime(at))"
        case .timetable: "Offline · timetable"
        case .unavailable: "Offline"
        }
    }

    var warns: Bool {
        switch self {
        case .updated, .lastKnown: false
        case .offline, .timetable, .unavailable: true
        }
    }
}

func widgetFreshness(_ board: BoardData, lead: Journey?) -> WidgetFreshness {
    if board.offline && board.generatedAt <= 0 { return .unavailable }
    if board.offline { return board.source == "schedule" ? .timetable : .offline(board.generatedAt) }
    return lead?.retained == true ? .lastKnown(board.generatedAt) : .updated(board.generatedAt)
}

func canonicalWidgetModes(_ modes: Set<String>) -> [String] {
    ["train", "metro", "ferry"].filter(modes.contains)
}

// Sydney's UTC offsets are whole hours, so its hour boundaries are UTC's.
func widgetHourStart(_ t: Millis) -> Millis {
    (t / hourMillis).rounded(.down) * hourMillis
}

func widgetSchedule(
    from now: Millis,
    hours: Int = widgetScheduleHours,
    predict: (Millis) -> (tripId: String, reverse: Bool)?
) -> [WidgetScheduleEntry] {
    let start = widgetHourStart(now)
    return (0..<hours).compactMap { hour in
        let at = start + Millis(hour) * hourMillis
        return predict(at).map { WidgetScheduleEntry(at: at, tripId: $0.tripId, reverse: $0.reverse) }
    }
}

private func weekHour(_ t: Millis) -> Int {
    let parts = widgetCalendar.dateComponents([.weekday, .hour], from: Date(timeIntervalSince1970: t / 1_000))
    return (parts.weekday ?? 0) * 24 + (parts.hour ?? 0)
}

func widgetScheduleEntry(_ schedule: [WidgetScheduleEntry], at t: Millis) -> WidgetScheduleEntry? {
    guard let first = schedule.first, t.isFinite else { return nil }
    let offset = ((t - first.at) / hourMillis).rounded(.down)
    if offset >= 0, offset < Millis(schedule.count) {
        let entry = schedule[Int(offset)]
        if entry.at <= t, t < entry.at + hourMillis { return entry }
    }
    // Local weekday and hour, not elapsed weeks, so a daylight-saving change keeps 8am at 8am.
    for back in 0..<3 {
        let key = weekHour(t - Millis(back) * hourMillis)
        if let match = schedule.last(where: { weekHour($0.at) == key }) { return match }
    }
    return nil
}

func widgetAnswer(_ snapshot: WidgetSnapshot, at t: Millis) -> WidgetAnswer? {
    if let focus = snapshot.focus, t <= focus.expiresAt,
       let trip = snapshot.trips.first(where: { $0.id == focus.tripId }) {
        return WidgetAnswer(trip: trip, reverse: focus.reverse, focus: focus)
    }
    guard let entry = widgetScheduleEntry(snapshot.schedule, at: t),
          let trip = snapshot.trips.first(where: { $0.id == entry.tripId }) else { return nil }
    return WidgetAnswer(trip: trip, reverse: entry.reverse)
}

func widgetRequest(for answer: WidgetAnswer, in snapshot: WidgetSnapshot, now: Millis) -> WidgetRequest {
    let pair = "\(answer.from.id)|\(answer.to.id)"
    if let focus = answer.focus {
        // The followed service is asked for whole, as the app does: every mode, no cap, its own departure window once gone.
        let departed = focus.journey.departure < now
        return WidgetRequest(
            key: "focus|\(pair)|\(focus.journey.key)",
            from: answer.from.station,
            to: answer.to.station,
            modes: allModes,
            at: departed ? max(now - 86_400_000, focus.journey.departure) : nil,
            focusJourneyKey: focus.journey.key,
            fallback: focus.board
        )
    }
    return WidgetRequest(
        key: "pair|\(pair)|\(snapshot.modes.joined(separator: ","))|\(snapshot.transferCap.map(String.init) ?? "")",
        from: answer.from.station,
        to: answer.to.station,
        modes: Set(snapshot.modes),
        transferLimit: snapshot.transferCap,
        fallback: snapshot.board(from: answer.from.id, to: answer.to.id)
    )
}

func widgetNextAnswerChange(_ snapshot: WidgetSnapshot, after t: Millis, until: Millis) -> Millis? {
    if let focus = snapshot.focus, t <= focus.expiresAt, snapshot.trips.contains(where: { $0.id == focus.tripId }) {
        return focus.expiresAt + 1 <= until ? focus.expiresAt + 1 : nil
    }
    let current = widgetAnswer(snapshot, at: t)
    var boundary = widgetHourStart(t) + hourMillis
    while boundary <= until {
        if widgetAnswer(snapshot, at: boundary) != current { return boundary }
        boundary += hourMillis
    }
    return nil
}

func widgetRequests(_ snapshot: WidgetSnapshot, from now: Millis, until: Millis, limit: Int = 3) -> [WidgetRequest] {
    var requests: [WidgetRequest] = []
    var t = now
    while requests.count < limit {
        if let answer = widgetAnswer(snapshot, at: t) {
            let request = widgetRequest(for: answer, in: snapshot, now: now)
            if !requests.contains(where: { $0.key == request.key }) { requests.append(request) }
        }
        guard let next = widgetNextAnswerChange(snapshot, after: t, until: until) else { break }
        t = next
    }
    return requests
}

/// A failed fetch shows the app's last board as the app shows a retained one; with none, the answer is offline.
func widgetSource(_ request: WidgetRequest, fetched: BoardData?) -> BoardData {
    if let fetched {
        guard let key = request.focusJourneyKey, !fetched.journeys.contains(where: { $0.key == key }),
              let fallback = request.fallback else { return fetched }
        // A working answer that no longer carries the followed service leaves its last copy last known, not offline.
        var kept = retainedOfflineBoard(fallback)
        kept.offline = false
        return kept
    }
    return request.fallback.map(retainedOfflineBoard)
        ?? BoardData(from: request.from, to: request.to, generatedAt: 0, offline: true)
}

func widgetContent(_ snapshot: WidgetSnapshot, sources: [String: BoardData], at t: Millis) -> WidgetContent {
    guard let answer = widgetAnswer(snapshot, at: t) else { return WidgetContent(date: t) }
    let board = sources[widgetRequest(for: answer, in: snapshot, now: t).key]
    let modes = Set(snapshot.modes)
    let upcoming = (board?.journeys ?? [])
        .filter { snapshot.eligible($0) && $0.effectiveDeparture >= t }
        .sorted { $0.effectiveDeparture < $1.effectiveDeparture }
    let candidates = board.map(recommendationCandidates) ?? []
    var lead: Journey?
    var replaced: Journey?
    if let focus = answer.focus {
        let followed = board?.journeys.first { $0.key == focus.journey.key } ?? focus.journey
        lead = followed
        if followed.cancelled, t < followed.effectiveDeparture,
           let replacement = selectRecommendation(
               candidates.filter { $0.journey.effectiveDeparture > followed.effectiveDeparture },
               now: t, modes: modes, maxTransfers: snapshot.transferCap
           ) {
            lead = replacement.journey
            replaced = followed
        }
    } else {
        lead = selectRecommendation(candidates, now: t, modes: modes, maxTransfers: snapshot.transferCap)?.journey
            ?? upcoming.first { !$0.cancelled } ?? upcoming.first
        replaced = upcoming.first.flatMap { $0.cancelled && $0.key != lead?.key ? $0 : nil }
    }
    var following: [Journey] = []
    if let lead, t <= lead.effectiveDeparture {
        let later = upcoming.firstIndex { $0.key == lead.key }.map { Array(upcoming[($0 + 1)...]) }
            ?? upcoming.filter { $0.effectiveDeparture > lead.effectiveDeparture }
        following = Array(later.filter { $0.key != replaced?.key }.prefix(2))
    }
    return WidgetContent(
        date: t,
        answer: answer,
        board: board,
        lead: lead,
        replaced: replaced,
        following: following,
        freshness: board.map { widgetFreshness($0, lead: lead) }
    )
}

/// Minute boundaries keep the printed countdown true; the exact service times move the lead the moment it leaves.
func widgetEntryDates(_ snapshot: WidgetSnapshot, sources: [String: BoardData], from now: Millis, until: Millis) -> [Millis] {
    var dates = Set<Millis>([now])
    var minute = (now / 60_000).rounded(.down) * 60_000 + 60_000
    while minute <= until {
        dates.insert(minute)
        minute += 60_000
    }
    var t = now
    while let change = widgetNextAnswerChange(snapshot, after: t, until: until) {
        dates.insert(change)
        t = change
    }
    for board in sources.values {
        for leg in board.journeys.flatMap(\.legs) {
            for instant in [leg.effectiveDeparture, leg.effectiveArrival] where instant > now && instant <= until {
                dates.insert(instant + 1)
            }
        }
    }
    return dates.sorted()
}

func widgetTimeline(
    _ snapshot: WidgetSnapshot,
    sources: [String: BoardData],
    from now: Millis,
    until: Millis,
    limit: Int = 300
) -> [WidgetContent] {
    widgetEntryDates(snapshot, sources: sources, from: now, until: until)
        .prefix(limit)
        .map { widgetContent(snapshot, sources: sources, at: $0) }
}

/// Board freshness alone never asks the system to redraw; the widget keeps its own refresh budget.
func widgetAnswerChanged(_ old: WidgetSnapshot?, _ new: WidgetSnapshot) -> Bool {
    guard let old else { return true }
    func focusIdentity(_ focus: WidgetFocus?) -> [String] {
        focus.map { [$0.tripId, String($0.reverse), String($0.pinned), $0.journey.key, String($0.expiresAt)] } ?? []
    }
    return old.trips != new.trips || old.schedule != new.schedule || old.modes != new.modes
        || old.transferCap != new.transferCap || focusIdentity(old.focus) != focusIdentity(new.focus)
}

func widgetContainerURL() -> URL? {
    FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: widgetAppGroup)
}

func readWidgetSnapshot(directory: URL?) -> WidgetSnapshot? {
    guard let url = directory?.appendingPathComponent(widgetSnapshotFileName),
          let data = try? Data(contentsOf: url),
          let snapshot = try? JSONDecoder().decode(WidgetSnapshot.self, from: data),
          snapshot.schemaVersion == 1 else { return nil }
    return snapshot
}

File: android/app/src/main/java/com/ilovetrains/app/WidgetSnapshot.kt
package com.ilovetrains.app

import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant

const val WidgetSnapshotFile = "widget-v1.json"
const val WidgetScheduleHours = 168
const val WidgetBoardLimit = 8
const val WidgetLiveRefresh = 15 * 60_000L
const val WidgetFollowingLimit = 4
// The chronometer counts below zero until a redraw runs, so a boundary this close is drawn as already passed.
const val WidgetRedrawLead = 15_000L
// JobScheduler holds a delayed job's alarm for up to three quarters of the delay (measured: 101 s late on a 136 s delay).
private const val JobLateness = 1.75
private const val RedrawMargin = 5_000L
private const val HourMillis = 3_600_000L

data class WidgetStop(val id: String, val name: String, val modes: List<String>) {
    val station get() = Station(id, name, modes = modes.toSet())
}
data class WidgetTrip(val id: String, val from: WidgetStop, val to: WidgetStop)
data class WidgetScheduleEntry(val at: Long, val tripId: String, val reverse: Boolean)
data class WidgetFocus(val tripId: String, val reverse: Boolean, val pinned: Boolean, val journey: Journey,
    val board: BoardData, val expiresAt: Long)

/** Everything the widget may know about the rider; only the app writes it. */
data class WidgetSnapshot(
    val writtenAt: Long,
    val trips: List<WidgetTrip>,
    val schedule: List<WidgetScheduleEntry>,
    val focus: WidgetFocus?,
    val modes: List<String>,
    val transferCap: Int?,
    val boards: List<BoardData>,
) {
    fun board(from: String, to: String) = boards.find { it.from.id == from && it.to.id == to }
    fun eligible(journey: Journey) =
        journey.legs.isNotEmpty() && journey.legs.all { it.mode in modes } && journey.withinTransferCap(transferCap)
}

data class WidgetAnswer(val trip: WidgetTrip, val reverse: Boolean, val focus: WidgetFocus? = null) {
    val from get() = if (reverse) trip.to else trip.from
    val to get() = if (reverse) trip.from else trip.to
}

data class WidgetRequest(
    val key: String,
    val from: Station,
    val to: Station,
    val modes: Set<String>,
    val at: Long?,
    val transferLimit: Int?,
    val focusJourneyKey: String? = null,
    val fallback: BoardData? = null,
)

/** [next] is the header's train; [cancelled] the struck service it replaced; [following] the board after it, in departure order. */
data class WidgetContent(
    val date: Long,
    val answer: WidgetAnswer?,
    val board: BoardData?,
    val next: Journey?,
    val following: List<Journey>,
    val provenance: String?,
    val cancelled: Journey? = null,
)

data class WidgetScheduleInputs(val trips: List<SavedTrip>, val history: List<ViewEvent>, val modes: Set<String>,
    val lastTripId: String?, val lastReverse: Boolean, val start: Long) {
    constructor(data: UserData, now: Long) : this(data.trips, data.history, data.modes, data.lastTripId, data.lastReverse,
        widgetHourStart(now))
}

fun canonicalWidgetModes(modes: Set<String>) = listOf("train", "metro", "ferry").filter { it in modes }

// Sydney's UTC offsets are whole hours, so its hour boundaries are UTC's.
fun widgetHourStart(t: Long) = Math.floorDiv(t, HourMillis) * HourMillis

fun widgetSchedule(from: Long, hours: Int = WidgetScheduleHours, predict: (Long) -> Pair<String, Boolean>?): List<WidgetScheduleEntry> {
    val start = widgetHourStart(from)
    return (0 until hours).mapNotNull { hour ->
        val at = start + hour * HourMillis
        predict(at)?.let { WidgetScheduleEntry(at, it.first, it.second) }
    }
}

fun widgetSchedule(data: UserData, stations: List<Station>, now: Long) =
    widgetSchedule(now) { at -> predict(data, stations, null, at)?.let { it.tripId to it.reverse } }

private fun weekHour(t: Long): Int = Instant.ofEpochMilli(t).atZone(Sydney).let { it.dayOfWeek.value * 24 + it.hour }

fun widgetScheduleEntry(schedule: List<WidgetScheduleEntry>, t: Long): WidgetScheduleEntry? {
    val first = schedule.firstOrNull() ?: return null
    val offset = Math.floorDiv(t - first.at, HourMillis)
    if (offset in schedule.indices) {
        val entry = schedule[offset.toInt()]
        if (t >= entry.at && t < entry.at + HourMillis) return entry
    }
    // Local weekday and hour, not elapsed weeks, so a daylight-saving change keeps 8am at 8am.
    for (back in 0 until 3) {
        val key = weekHour(t - back * HourMillis)
        schedule.lastOrNull { weekHour(it.at) == key }?.let { return it }
    }
    return null
}

fun widgetAnswer(snapshot: WidgetSnapshot, t: Long): WidgetAnswer? {
    snapshot.focus?.takeIf { t <= it.expiresAt }?.let { focus ->
        snapshot.trips.find { it.id == focus.tripId }?.let { return WidgetAnswer(it, focus.reverse, focus) }
    }
    val entry = widgetScheduleEntry(snapshot.schedule, t) ?: return null
    return snapshot.trips.find { it.id == entry.tripId }?.let { WidgetAnswer(it, entry.reverse) }
}

fun widgetRequest(answer: WidgetAnswer, snapshot: WidgetSnapshot, now: Long): WidgetRequest {
    val pair = "${answer.from.id}|${answer.to.id}"
    val focus = answer.focus
    if (focus != null) {
        // The followed service is asked for whole, as the app does: every mode, no cap, its own departure window once gone.
        val departed = focus.journey.departure < now
        return WidgetRequest("focus|$pair|${focus.journey.key}", answer.from.station, answer.to.station, AllModes,
            if (departed) maxOf(now - 86_400_000, focus.journey.departure) else null, null, focus.journey.key, focus.board)
    }
    return WidgetRequest("pair|$pair|${snapshot.modes.joinToString(",")}|${snapshot.transferCap ?: ""}",
        answer.from.station, answer.to.station, snapshot.modes.toSet(), null, snapshot.transferCap,
        fallback = snapshot.board(answer.from.id, answer.to.id))
}

fun widgetNextAnswerChange(snapshot: WidgetSnapshot, after: Long, until: Long): Long? {
    val focus = snapshot.focus
    if (focus != null && after <= focus.expiresAt && snapshot.trips.any { it.id == focus.tripId }) {
        return (focus.expiresAt + 1).takeIf { it <= until }
    }
    val current = widgetAnswer(snapshot, after)
    var boundary = widgetHourStart(after) + HourMillis
    while (boundary <= until) {
        if (widgetAnswer(snapshot, boundary) != current) return boundary
        boundary += HourMillis
    }
    return null
}

/** A failed or unmatched fetch shows the app's last board as the app shows a retained one. */
fun widgetSource(request: WidgetRequest, fetched: BoardData?): BoardData? {
    if (fetched != null) {
        val key = request.focusJourneyKey ?: return fetched
        if (fetched.journeys.any { it.key == key }) return fetched
    }
    return request.fallback?.lastKnown()
}

/** The tracker's wording, never LIVE: a widget's data is minutes old for nearly all of its life. */
fun widgetFreshness(board: BoardData): String = when {
    board.offline && board.source != "live" -> "Offline · timetable"
    board.offline -> "Offline · Last updated ${clockTime(board.generatedAt)}"
    else -> "Last updated ${clockTime(board.generatedAt)}"
}

internal fun widgetFocused(focus: WidgetFocus, board: BoardData?) = FocusedJourney(focus.tripId, focus.reverse,
    board?.journeys?.find { it.key == focus.journey.key } ?: focus.journey, board ?: focus.board, focus.pinned)

fun widgetContent(snapshot: WidgetSnapshot, sources: Map<String, BoardData>, t: Long): WidgetContent {
    val answer = widgetAnswer(snapshot, t) ?: return WidgetContent(t, null, null, null, emptyList(), null)
    val board = sources[widgetRequest(answer, snapshot, t).key]
    val provenance = board?.let(::widgetFreshness) ?: "Offline"
    // A service leaving at the redraw instant has gone for a view that cannot redraw each minute.
    val lead = homeAnswer(board, answer.focus?.let { widgetFocused(it, board) }, t + 1, snapshot.modes.toSet(), snapshot.transferCap)
        ?: return WidgetContent(t, answer, board, null, emptyList(), provenance)
    val following = board?.journeys.orEmpty().filter {
        snapshot.eligible(it) && it.key != lead.journey.key && it.key != lead.cancelledLead?.key &&
            it.effectiveDeparture > maxOf(t, lead.journey.effectiveDeparture - 1)
    }.sortedBy { it.effectiveDeparture }.take(WidgetFollowingLimit)
    return WidgetContent(t, answer, board, lead.journey, following, provenance, lead.cancelledLead)
}

data class WidgetDraw(val at: Long, val redraw: Long?)

/**
 * The next boundary is approached in redraws that each land before it however late the job runs, and one within
 * [WidgetRedrawLead] draws the widget as it stands after the boundary, so a countdown never passes zero.
 */
fun widgetDraw(snapshot: WidgetSnapshot, sources: Map<String, BoardData>, now: Long, until: Long): WidgetDraw {
    var at = now
    while (true) {
        val next = widgetNextBoundary(snapshot, sources, at, until) ?: return WidgetDraw(at, null)
        if (next - now > WidgetRedrawLead) return WidgetDraw(at, now + ((next - now - RedrawMargin) / JobLateness).toLong())
        at = next
    }
}

fun widgetNextBoundary(snapshot: WidgetSnapshot, sources: Map<String, BoardData>, after: Long, until: Long): Long? {
    val candidates = mutableListOf<Long>()
    widgetNextAnswerChange(snapshot, after, until)?.let(candidates::add)
    val answer = widgetAnswer(snapshot, after)
    val board = answer?.let { sources[widgetRequest(it, snapshot, after).key] }
    if (answer != null && board != null) {
        val times = answer.focus?.let { focus -> widgetFocused(focus, board).journey.legs.flatMap { listOf(it.effectiveDeparture, it.effectiveArrival) } }
            ?: board.journeys.filter(snapshot::eligible).map { it.effectiveDeparture }
        times.filter { it > after }.minOrNull()?.let(candidates::add)
        if (board.isLive(after)) candidates.add(board.generatedAt + 90_001)
    }
    return candidates.filter { it in (after + 1)..until }.minOrNull()
}

/** Board freshness alone never asks the launcher to redraw; the widget keeps its own refresh budget. */
fun widgetAnswerChanged(old: WidgetSnapshot?, new: WidgetSnapshot): Boolean {
    if (old == null) return true
    fun identity(focus: WidgetFocus?) = focus?.let { listOf(it.tripId, it.reverse, it.pinned, it.journey.key, it.expiresAt) }
    return old.trips != new.trips || old.schedule != new.schedule || old.modes != new.modes ||
        old.transferCap != new.transferCap || identity(old.focus) != identity(new.focus)
}

fun widgetBoardPairs(data: UserData, schedule: List<WidgetScheduleEntry>): List<Pair<Station, Station>> =
    schedule.mapNotNull { entry ->
        data.trips.find { it.id == entry.tripId }?.let { if (entry.reverse) it.to to it.from else it.from to it.to }
    }.distinctBy { "${it.first.id}|${it.second.id}" }

fun widgetSnapshot(data: UserData, schedule: List<WidgetScheduleEntry>, boards: List<BoardData>, now: Long): WidgetSnapshot {
    fun stop(station: Station) = WidgetStop(station.id, station.name, canonicalWidgetModes(station.modes))
    fun slim(board: BoardData, journeys: List<Journey>) = BoardData(board.from, board.to, journeys, board.generatedAt,
        board.source, board.offline, board.serverStale)
    val trips = data.trips.filter { compatible(it, data.modes) }.map { WidgetTrip(it.id, stop(it.from), stop(it.to)) }
    val pairs = widgetBoardPairs(data, schedule).map { "${it.first.id}|${it.second.id}" }.toSet()
    val published = boards.filter { "${it.from.id}|${it.to.id}" in pairs }.distinctBy { "${it.from.id}|${it.to.id}" }
        .map { board ->
            slim(board, board.journeys.filter {
                it.legs.all { leg -> leg.mode in data.modes } && it.withinTransferCap(data.maxTransfers) && it.effectiveDeparture >= now
            }.take(WidgetBoardLimit))
        }
    val focus = visibleFocus(data, now)?.takeIf { it.journey.withinTransferCap(data.maxTransfers) }?.let {
        WidgetFocus(it.tripId, it.reverse, it.pinned, it.journey, slim(it.board, listOf(it.journey)), focusExpiry(it))
    }
    return WidgetSnapshot(now, trips, if (trips.isEmpty()) emptyList() else schedule, focus,
        canonicalWidgetModes(data.modes), data.maxTransfers, published)
}

object WidgetWire {
    private fun stop(s: WidgetStop) = JSONObject().put("id", s.id).put("name", s.name).put("modes", JSONArray(s.modes))
    private fun stop(o: JSONObject) = WidgetStop(o.getString("id"), o.getString("name"),
        o.getJSONArray("modes").let { a -> (0 until a.length()).map { a.getString(it) } })
    private fun trip(t: WidgetTrip) = JSONObject().put("id", t.id).put("from", stop(t.from)).put("to", stop(t.to))
    private fun trip(o: JSONObject) = WidgetTrip(o.getString("id"), stop(o.getJSONObject("from")), stop(o.getJSONObject("to")))
    private fun focus(f: WidgetFocus) = JSONObject().put("tripId", f.tripId).put("reverse", f.reverse).put("pinned", f.pinned)
        .put("journey", Wire.journey(f.journey)).put("board", Wire.board(f.board)).put("expiresAt", f.expiresAt)
    private fun focus(o: JSONObject) = WidgetFocus(o.getString("tripId"), o.getBoolean("reverse"), o.getBoolean("pinned"),
        Wire.journey(o.getJSONObject("journey")), Wire.board(o.getJSONObject("board")), o.getLong("expiresAt"))

    fun snapshot(s: WidgetSnapshot): JSONObject = JSONObject().put("schemaVersion", 1).put("writtenAt", s.writtenAt)
        .put("trips", s.trips.jsonEach(::trip))
        .put("schedule", s.schedule.jsonEach { JSONObject().put("at", it.at).put("tripId", it.tripId).put("reverse", it.reverse) })
        .put("focus", s.focus?.let(::focus)).put("modes", JSONArray(s.modes)).put("transferCap", s.transferCap)
        .put("boards", s.boards.jsonEach(Wire::board))

    fun snapshot(o: JSONObject): WidgetSnapshot {
        require(o.getInt("schemaVersion") == 1)
        return WidgetSnapshot(o.getLong("writtenAt"), o.getJSONArray("trips").readEach(::trip),
            o.getJSONArray("schedule").readEach { WidgetScheduleEntry(it.getLong("at"), it.getString("tripId"), it.getBoolean("reverse")) },
            o.optJSONObject("focus")?.let { runCatching { focus(it) }.getOrNull() },
            o.getJSONArray("modes").let { a -> (0 until a.length()).map { a.getString(it) } },
            if (o.has("transferCap")) o.getInt("transferCap") else null,
            o.getJSONArray("boards").readEach(Wire::board))
    }

    fun content(c: WidgetContent): JSONObject = JSONObject().put("date", c.date)
        .put("answer", c.answer?.let { a -> JSONObject().put("trip", trip(a.trip)).put("reverse", a.reverse).put("focus", a.focus?.let(::focus)) })
        .put("board", c.board?.let(Wire::board)).put("next", c.next?.let(Wire::journey))
        .put("following", c.following.jsonEach(Wire::journey)).put("provenance", c.provenance)
        .put("cancelled", c.cancelled?.let(Wire::journey))

    fun content(o: JSONObject) = WidgetContent(o.getLong("date"),
        o.optJSONObject("answer")?.let { a -> WidgetAnswer(trip(a.getJSONObject("trip")), a.getBoolean("reverse"),
            a.optJSONObject("focus")?.let(::focus)) },
        o.optJSONObject("board")?.let(Wire::board), o.optJSONObject("next")?.let(Wire::journey),
        o.getJSONArray("following").readEach(Wire::journey), o.stringOrNull("provenance"),
        o.optJSONObject("cancelled")?.let(Wire::journey))
}
